src/main/java/com/conveyal/datatools/manager/models/FeedSource.java

```diff
@@ -479,77 +459,101 @@ public String latestVersionId() {
-            if (feedVersion != null) {
-                return feedVersion.validationSummary().startDate;
-            }
-        }
-        return deployedFeedVersionStartDate;
+        FeedVersion feedVersion = getDeployedFeedVersion();
+        return feedVersion != null ? feedVersion.validationSummary().startDate : null;
     }
 
     @JsonInclude(JsonInclude.Include.NON_NULL)
     @JsonView(JsonViews.UserInterface.class)
     @JsonProperty("deployedFeedVersionEndDate")
     public LocalDate getDeployedFeedVersionEndDate() {
-        if (deployedFeedVersionStartDate == null) {
-            FeedVersion feedVersion = getDeployedFeedVersionViaExpensiveQuery();
-            if (feedVersion != null) {
-                return feedVersion.validationSummary().startDate;
-            }
-        }
-        return deployedFeedVersionEndDate;
+        FeedVersion feedVersion = getDeployedFeedVersion();
+        return feedVersion != null ? feedVersion.validationSummary().endDate : null;
     }
 
     /**
-     * Find the latest deployment containing a feed version for a feed source. This method uses expensive mongo queries instead
-     * of relying on the deployment job to update the static fields. However, it must be kept around to accurately
-     * display feeds that haven't been deployed via the new deploy job.
+     * The deployed feed version.
+     * This cannot be returned because of a circular reference between feed source and feed version. Instead, individual
+     * parameters (version id, start date and end date) are returned.
+     */
+    @JsonIgnore
+    @BsonIgnore
+    private FeedVersion deployedFeedVersion = null;
+
+    /**
+     * If a project has a "pinned" deployment it's feed versions take precedence over the latest feed version for this
+     * feed source. In this case, return either the latest "pinned" feed version or just the feed version, which ever
+     * is available.
      *
-     * When the static fields are present this is not needed.
+     * If a project does not have a "pinned" deployment (or the above is not available), return the latest feed version
+     * for this feed source from the newest deployment.
      */
-    private FeedVersion getDeployedFeedVersionViaExpensiveQuery() {
+    private FeedVersion getDeployedFeedVersion() {
         if (deployedFeedVersion == null) {
-            Collection<Deployment> deployments = Persistence.deployments.getFiltered(eq("projectId", this.projectId), Sorts.descending("lastUpdated"));
-            Collection<FeedVersion> feedVersions = Persistence.feedVersions.getFiltered(eq("feedSourceId", this.id), Sorts.descending("updated"));
-            if (deployments.isEmpty() || feedVersions.isEmpty()) {
-                return null;
+            Project project = Persistence.projects.getById(this.projectId);
+            if (project.pinnedDeploymentId != null) {
+                Deployment deployment = Persistence.deployments.getById(project.pinnedDeploymentId);
```

Comment:
I believe this line is causing the slowdown. Would it be possible to cache the deployments in memory on startup and try searching that cache before resorting to this?

Reply:
I have made some further updates to reduce objects being returned and in the case of deployment objects limiting the fields returned. I'm hoping this is enough to avoid caching as managing that will be quite onerous.